compiler/rustc_mir/src/transform/flatten_locals.rs

```diff
@@ -0,0 +1,281 @@
+                        user_ty: None,
+                        ..self.local_decls[place.local].clone()
+                    });
+                    v.insert(local);
+                }
+            }
+        }
+    }
+
+    impl<'tcx, 'll> Visitor<'tcx> for PreFlattenVisitor<'tcx, 'll> {
+        fn visit_statement(&mut self, statement: &Statement<'tcx>, location: Location) {
+            if let StatementKind::SetDiscriminant { place, .. } = &statement.kind {
+                if place.projection.is_empty() {
+                    if self.create_discriminant(place.local) {
+                        return;
+                    }
+                }
+            }
+            self.super_statement(statement, location)
+        }
+
+        fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+            if let Rvalue::Discriminant(place) = rvalue {
+                if let Some(local) = place.as_local() {
+                    if self.create_discriminant(local) {
+                        return;
+                    }
+                }
+            }
+            self.super_rvalue(rvalue, location)
+        }
+
+        fn visit_place(&mut self, place: &Place<'tcx>, _: PlaceContext, _: Location) {
+            let nproj = match &place.projection[..] {
+                &[PlaceElem::Field(..), ..] => 1,
+                &[PlaceElem::Downcast(..), PlaceElem::Field(..), ..] => 2,
+                _ => return,
+            };
+            let pr = PlaceRef { local: place.local, projection: &place.projection[..nproj] };
+            self.create_place(pr)
+        }
+    }
+}
+
+struct ReplacementVisitor<'tcx, 'll> {
+    tcx: TyCtxt<'tcx>,
+    map: &'ll ReplacementMap<'tcx>,
+    all_dead_locals: FxHashSet<Local>,
+}
+
+impl<'tcx, 'll> MutVisitor<'tcx> for ReplacementVisitor<'tcx, 'll> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_statement(&mut self, statement: &mut Statement<'tcx>, location: Location) {
+        match &statement.kind {
+            StatementKind::SetDiscriminant { place, variant_index }
+                if place.projection.is_empty() =>
+            {
+                if let Some((local, discr_ty)) = self.map.discr.get(&place.local) {
+                    statement.kind = StatementKind::Assign(box (
+                        Place::from(*local),
+                        Rvalue::Use(Operand::const_from_scalar(
+                            self.tcx,
+                            discr_ty,
+                            Scalar::from_u64(variant_index.as_u32().into()),
+                            statement.source_info.span,
+                        )),
+                    ));
+
+                    return;
+                }
+            }
+            StatementKind::StorageLive(local) | StatementKind::StorageDead(local)
+                if self.all_dead_locals.contains(local) =>
+            {
+                statement.make_nop();
+                return;
+            }
```

Comment:
I think we should emit storage markers for all flattened fields instead of completely removing them completely.